torch/distributed/tensor/_api.py

```diff
@@ -212,6 +221,244 @@ def backward(ctx, grad_output: "DTensor"):  # type: ignore[override]
+
+    # set default placements to replicated if not specified
+    placement_tuple: tuple[Placement, ...]
+    if placements is None:
+        assert shard_order is not None
+        # convert from shard_order to placements
+        placement_tuple = _convert_shard_order_to_placements(
+            normalized_shard_order, device_mesh
+        )
+        shard_order_tuple = _from_dict_to_ShardOrder(normalized_shard_order)
+    else:
+        normalized_placements = list(placements)  # type: ignore[assignment]
+        for i, placement in enumerate(placements):
+            if placement.is_shard():
+                tensor_dim = placement.dim  # type: ignore[attr-defined]
+                if tensor_dim < -tensor_rank or tensor_dim >= tensor_rank:
+                    raise ValueError(
+                        f"tensor dim {tensor_dim} specified in `placements` is out of range for tensor_rank {tensor_rank}."
+                    )
+                tensor_dim = tensor_dim + tensor_rank if tensor_dim < 0 else tensor_dim
+
+                # reconstruct `placement` object in case it is `_StridedShard` for backward compatibility
+                if isinstance(placement, _StridedShard):
+                    has_stridedshard_in_placements = True
+                    normalized_placements[i] = _StridedShard(
+                        tensor_dim, split_factor=placement.split_factor
+                    )
+                else:
+                    assert type(placement) is Shard, (
+                        "Expected placement to be exactly of type Shard"
+                    )
+                    normalized_placements[i] = Shard(tensor_dim)
+
+        placement_tuple = tuple(normalized_placements)
+
+        if shard_order is None:
+            shard_order_tuple = DTensorSpec.compute_default_shard_order(placement_tuple)
+        else:
+            # both shard_order and placements are specified; need to validate their correctness
+            if has_stridedshard_in_placements:
+                # _StridedShard doesn't work with shard_order
+                raise RuntimeError(
+                    "Cannot specify both `placements` and `shard_order` when "
+                    "`placements` contains `_StridedShard`!"
+                )
+
+            # Build map from tensor_dim -> list of mesh_dims (in left-to-right order)
+            tensor_dim_to_mesh_dims: dict[int, list[int]] = defaultdict(list)
+            for mesh_dim, placement in enumerate(placement_tuple):
+                if placement.is_shard():
+                    tensor_dim = placement.dim  # type: ignore[attr-defined]
+                    tensor_dim_to_mesh_dims[tensor_dim].append(mesh_dim)
+
+            # Validate: if a tensor_dim is in shard_order, all its sharding mesh dims must be listed
+            for tensor_dim, mesh_dims_in_order in normalized_shard_order.items():
+                expected = set(tensor_dim_to_mesh_dims.get(tensor_dim, []))
```

Comment:
should we use set here? wouldn't it be better to check that the specified order matches between the two?